envoy/network/filter.h

```diff
@@ -647,6 +647,217 @@ class UdpListenerFilterManager {
 using UdpListenerFilterFactoryCb = std::function<void(
     UdpListenerFilterManager& udp_listener_filter_manager, UdpReadFilterCallbacks& callbacks)>;
 
+/**
+ * Common interface for ReadFilterCallbacks and WriteFilterCallbacks.
+ */
+class UdpSessionFilterCallbacks {
+public:
+  virtual ~UdpSessionFilterCallbacks() = default;
+
+  /**
+   * @return uint64_t the ID of the originating UDP session.
+   */
+  virtual uint64_t sessionId() const PURE;
+
+  /**
+   * @return StreamInfo for logging purposes.
+   */
+  virtual StreamInfo::StreamInfo& streamInfo() PURE;
+
+  /**
+   * Allows a filter to inject a datagram to successive filters in the session filter chain.
+   * The injected datagram will be iterated as a regular received datagram, and may also be
+   * stopped by further filters. This can be used, for example, to continue processing previously
+   * buffered datagrams by a filter after an asynchronous operation ended.
+   */
+  virtual void injectDatagramToFilterChain(Network::UdpRecvData& data) PURE;
+};
+
+class UdpSessionReadFilterCallbacks : public UdpSessionFilterCallbacks {
+public:
+  ~UdpSessionReadFilterCallbacks() override = default;
+
+  /**
+   * If a read filter stopped filter iteration, continueFilterChain() can be called to continue the
+   * filter chain. It will have onNewSession() called if it was not previously called.
+   * @return false if the session is removed and no longer valid, otherwise returns true.
+   */
+  virtual bool continueFilterChain() PURE;
+};
+
+class UdpSessionWriteFilterCallbacks : public UdpSessionFilterCallbacks {};
+
+class UdpSessionFilterBase {
```

Comment:
It looks like this class was added as a base interface for both ``UdpSessionReadFilter`` and ``UdpSessionWriteFilter``, since the ``onSessionComplete()`` is common to both. It also provides a default implementation for that method (if I'm trying to do some sort or comparing with other languages). The base is not meant to be instantiated directly.

Anyways - this is a cut and paste from the file which this is moved from, if there's a need to refactor this, can we do it in a separate PR? I'd like not to modify logic in this one for the integrity of changes